File: build_site_data.py
from pathlib import Path
import csv
import json
import math
import statistics
import zipfile
from collections import defaultdict

ROOT = Path(__file__).parent
ZIP_PATH = ROOT / "Elite Prospects Hockey Stats & Player Data.zip"
JSON_PATH = ROOT / "nhl_player_data.json"
HISTOGRAM_PATH = ROOT / "assets" / "histogram.svg"


def normalize_position(pos: str) -> str:
    value = (pos or "").strip().upper()
    if value in {"C", "LW", "RW", "W", "F", "L", "R", "LF", "RF"}:
        return "Forward"
    if value in {"D", "LD", "RD", "DD"}:
        return "Defense"
    if value in {"G", "GK", "GOALIE"}:
        return "Goalie"
    return "Other"


def build_svg(histogram: list[dict], title: str) -> str:
    width = 800
    height = 430
    margin = {"left": 70, "right": 30, "top": 40, "bottom": 70}
    chart_w = width - margin["left"] - margin["right"]
    chart_h = height - margin["top"] - margin["bottom"]
    max_count = max(item["count"] for item in histogram)
    bar_gap = 12
    bar_w = (chart_w - (len(histogram) - 1) * bar_gap) / len(histogram)

    bars = []
    for index, item in enumerate(histogram):
        value = item["count"]
        x = margin["left"] + index * (bar_w + bar_gap)
        h = (value / max_count) * chart_h if max_count else 0
        y = margin["top"] + chart_h - h
        bars.append(
            f'<rect x="{x:.1f}" y="{y:.1f}" width="{bar_w:.1f}" height="{h:.1f}" rx="10" fill="#4fd1c5" opacity="0.95"/>'
        )
        bars.append(
            f'<text x="{x + bar_w / 2:.1f}" y="{height - 28}" text-anchor="middle" font-size="13" fill="#e6f7ff">{item["bin"]}</text>'
        )
        bars.append(
            f'<text x="{x + bar_w / 2:.1f}" y="{y - 8:.1f}" text-anchor="middle" font-size="12" fill="#f8fbff">{value}</text>'
        )

    axes = [
        f'<line x1="{margin["left"]}" y1="{margin["top"] + chart_h}" x2="{margin["left"]}" y2="{margin["top"]}" stroke="#e6f7ff" stroke-width="2"/>',
        f'<line x1="{margin["left"]}" y1="{margin["top"] + chart_h}" x2="{width - margin["right"]}" y2="{margin["top"] + chart_h}" stroke="#e6f7ff" stroke-width="2"/>',
        f'<text x="{width/2:.1f}" y="24" text-anchor="middle" font-size="24" font-weight="600" fill="#f8fbff">{title}</text>',
    ]

    return f'''<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}" role="img" aria-label="Histogram of NHL player heights">
  <rect width="{width}" height="{height}" rx="24" fill="#071b2f"/>
  {''.join(axes)}
  {''.join(bars)}
</svg>'''


def main() -> None:
    with zipfile.ZipFile(ZIP_PATH) as archive:
        dim_rows = csv.DictReader(archive.read("player_dim.csv").decode("latin-1").splitlines())
        stats_rows = csv.DictReader(archive.read("player_stats.csv").decode("latin-1").splitlines())

    players = {row["PLAYER_ID"]: row for row in dim_rows}
    for row in stats_rows:
        pid = row["PLAYER_ID"]
        if pid in players:
            players[pid].update(row)

    records = []
    for row in players.values():
        try:
            height = float(row.get("HEIGHT_CM", ""))
            if math.isfinite(height):
                records.append({
                    "height": height,
                    "position": normalize_position(row.get("PRIMARY_POS", "")),
                })
        except Exception:
            continue

    by_position = defaultdict(list)
    for record in records:
        by_position[record["position"]].append(record["height"])

    summary = []
    for position, values in sorted(by_position.items()):
        values = sorted(values)
        summary.append({
            "position": position,
            "count": len(values),
            "mean": round(statistics.mean(values), 1),
            "median": round(statistics.median(values), 1),
            "std": round(statistics.stdev(values), 1) if len(values) > 1 else 0.0,
        })

    bins = list(range(150, 211, 10))
    counts = [0] * (len(bins) - 1)
    for height in [record["height"] for record in records]:
        for idx in range(len(bins) - 1):
            if bins[idx] <= height < bins[idx + 1]:
                counts[idx] += 1
                break
        else:
            if height >= bins[-1]:
                counts[-1] += 1

    histogram = [{"bin": f"{bins[idx]}-{bins[idx + 1] - 1} cm", "count": counts[idx]} for idx in range(len(counts))]

    payload = {"summary": summary, "histogram": histogram, "records": records[:12]}
    JSON_PATH.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    HISTOGRAM_PATH.write_text(build_svg(histogram, "Distribution of NHL player heights"), encoding="utf-8")
    print(f"Wrote {JSON_PATH} and {HISTOGRAM_PATH}")
    print(json.dumps(summary, indent=2))


if __name__ == "__main__":
    main()

File: build_site_data.R
library(readr)
library(dplyr)
library(ggplot2)
library(stringr)
library(jsonlite)
library(purrr)

root <- getwd()
zip_path <- file.path(root, "Elite Prospects Hockey Stats & Player Data.zip")

extract_csv <- function(path, archive_path) {
  temp_dir <- tempfile("nhl_data")
  dir.create(temp_dir)
  unzip(zip_path, files = archive_path, exdir = temp_dir)
  read_csv(file.path(temp_dir, basename(archive_path)), show_col_types = FALSE)
}

player_dim <- extract_csv(zip_path, "player_dim.csv")
player_stats <- extract_csv(zip_path, "player_stats.csv")

position_lookup <- player_stats %>%
  filter(!is.na(PRIMARY_POS) & PRIMARY_POS != "") %>%
  mutate(PRIMARY_POS = str_to_upper(PRIMARY_POS)) %>%
  group_by(PLAYER_ID) %>%
  summarise(PRIMARY_POS = first(PRIMARY_POS), .groups = "drop")

players <- player_dim %>%
  distinct(PLAYER_ID, .keep_all = TRUE) %>%
  left_join(position_lookup, by = "PLAYER_ID") %>%
  mutate(
    height_cm = as.numeric(HEIGHT_CM)
  ) %>%
  filter(!is.na(height_cm), !is.na(PRIMARY_POS), PRIMARY_POS != "")

players <- players %>%
  mutate(
    position_group = case_when(
      PRIMARY_POS %in% c("C", "LW", "RW", "W", "F", "L", "R", "LF", "RF") ~ "Forward",
      PRIMARY_POS %in% c("D", "LD", "RD", "DD") ~ "Defense",
      PRIMARY_POS %in% c("G", "GK", "GOALIE") ~ "Goalie",
      TRUE ~ "Other"
    )
  )

summary_tbl <- players %>%
  group_by(position_group) %>%
  summarise(
    count = n(),
    mean_height = round(mean(height_cm, na.rm = TRUE), 1),
    median_height = round(median(height_cm, na.rm = TRUE), 1),
    sd_height = round(sd(height_cm, na.rm = TRUE), 1),
    .groups = "drop"
  )

hist_data <- data.frame(height_cm = players$height_cm)

hist_plot <- ggplot(hist_data, aes(x = height_cm)) +
  geom_histogram(binwidth = 5, fill = "#4fd1c5", color = "#071b2f", alpha = 0.95, linewidth = 0.5) +
  geom_density(aes(y = after_stat(count) * 5), color = "#ffd166", linewidth = 1.1, alpha = 0.7) +
  labs(
    title = "Distribution of NHL player heights",
    x = "Height (cm)",
    y = "Count"
  ) +
  xlim(150, 210) +
  theme_minimal(base_size = 16) +
  theme(
    plot.title = element_text(size = 18, face = "bold", color = "#f8fbff"),
    axis.title = element_text(color = "#f8fbff"),
    axis.text = element_text(color = "#f8fbff"),
    panel.background = element_rect(fill = "#071b2f", color = NA),
    plot.background = element_rect(fill = "#071b2f", color = NA),
    panel.grid.major = element_line(color = "#214a67"),
    panel.grid.minor = element_blank()
  )

box_plot <- ggplot(players, aes(x = position_group, y = height_cm, fill = position_group)) +
  geom_boxplot(alpha = 0.9, outlier.alpha = 0.4, width = 0.6) +
  geom_jitter(width = 0.12, alpha = 0.05, color = "#f8fbff") +
  scale_fill_manual(values = c("Defense" = "#4fd1c5", "Forward" = "#70b7ff", "Goalie" = "#ff6b6b", "Other" = "#f7d774")) +
  labs(
    title = "Height by position group",
    x = "Position group",
    y = "Height (cm)"
  ) +
  ylim(150, 210) +
  theme_minimal(base_size = 16) +
  theme(
    plot.title = element_text(size = 18, face = "bold", color = "#f8fbff"),
    axis.title = element_text(color = "#f8fbff"),
    axis.text = element_text(color = "#f8fbff"),
    panel.background = element_rect(fill = "#071b2f", color = NA),
    plot.background = element_rect(fill = "#071b2f", color = NA),
    panel.grid.major = element_line(color = "#214a67"),
    panel.grid.minor = element_blank(),
    legend.position = "none"
  )

# Export images
png("assets/histogram.png", width = 1000, height = 650, bg = "#071b2f")
print(hist_plot)
dev.off()

png("assets/boxplot.png", width = 1000, height = 650, bg = "#071b2f")
print(box_plot)
dev.off()

# Save summary data
write_json(summary_tbl, "nhl_player_data.json", pretty = TRUE, auto_unbox = TRUE)

cat("Generated charts and summary data.\n")
print(summary_tbl)
